### DISCUSSION TITLE
backup.velero.io/backup-volumes but for CSI snapshots

### DISCUSSION THREAD

**Discussion by makarov-roman** [vmware-tanzu/velero#8356] [Community support Q&A] [answered]
Hi,
currently we are supporting on-premise and cloud environments.

We're using FSB on-premise and switching to CSI snapshots on cloud. Backups were configured using `backup.velero.io/backup-volumes` annotation. 
Is there a way to achieve the same(opt-in) behavior with CSI snapshots?

Thanks in advance.

**Comment by kaovilai** [comment, marked as answer]
Yes. See https://velero.io/docs/v1.15/resource-filtering/#resource-policies

**Reply by kaovilai** [reply]
Actually if you want to filter by names perhaps not yet.

**Reply by kaovilai** [reply]
Opened: https://github.com/vmware-tanzu/velero/issues/8363

**Reply by makarov-roman** [reply]
@kaovilai thanks!
I think I would like to have https://github.com/vmware-tanzu/velero/issues/8256 instead. Because I want to keep this configuration (whether to backup volume or not) as part of application chart. 

If I'm not mistaken - resource policy, on the other hand, is a cluster wide resource and it will create unwanted dependencies for a deployments.

**Reply by kaovilai** [reply]
both #8363 and #8256 are via the same resource policy which would defines volume name for the whole cluster yes.

**Reply by kaovilai** [reply]
can you walk us through how you would want to annotate the object that needs to backup?

would you be annotating each volume with something like `backup.velero.io/snapshot: csi`?

**Reply by makarov-roman** [reply]
@kaovilai yes, it would be perfect for us. We're using standardised helm library across applications, so adding these annotations is a low effort, and they already present for FSB.

**Reply by sseago** [reply]
@makarov-roman But note that these annotations function to choose between fs-backup and snapshot, not to choose between fs-backup and not backing up. So the current "opt-out" approach (fs-backup-excludes) is the same thing as opt-in for snapshot. In other words, if you're defaulting to fs-backup, then any volume with "backup-volumes-excludes" set would use snapshots.

If you are not mixing snapshot and fs-backup in the same backup, then these annotations aren't really relevant for you, since these annotations don't include/exclude from backup, they merely tell velero whether to use snapshots vs. fs-backup. If you want to exclude volumes from any sort of content backup, you want the "skip" volume policy.

**Reply by kaovilai** [reply]
@sseago I sense that they want opt-in approach without defaulting to fs-backup via annotations. Not mixing, just CSI. Opting in via annotation.

**Reply by kaovilai** [reply]
Basically like [opt in via pod annotation](https://velero.io/docs/v1.15/file-system-backup/#using-opt-in-pod-volume-backup). But since here csi do not require pod, annotation will be on PVCs themselves.

**Reply by kaovilai** [reply]
IIUC velero may not be listing every resource type and parsing all things not yet marked as includedResources.
So it may be easier to "snapshot all volumes found in backup" then annotate to opt out.

Alternatively velero will need another feature flag to always list and parse each PVC to detect this "opt-in annotation".

**Reply by sseago** [reply]
@kaovilai right. But I think this is because there's basic confusion on what opt-in/opt-out pod annotations do for fs-backup. They're not opting in/out of backing up (i.e. fs-backup vs nothing) but opting in/out of fs-backup (i.e. fs-backup vs. snapshot) therefore opting out of fs-backup *is* opting in to snapshot backup. If they want to exclude certain volumes from the backup and use csi for the others, that is supported via volume policies. Once we implement the label selector bit, that means that they get exactly what is being asked here, but with PVC labels rather than pod annotations.

**Reply by kaovilai** [reply]
ok I think label selector is most reasonable. cause you can list via labelselector and/or name but definitely not via annotation (require listing all then looking for each object IIUC)

**Reply by sseago** [reply]
@kaovilai also using label selector with the existing volume policies functionality gives you the flexibility for both opt-in and opt-out. i.e. a policy that says "back up via snapshot with this label" followed by a generic "skip" policy, alternatively, you could reverse it and "skip PVCs with this label" followed by a generic "snapshot" policy.

**Comment by sseago** [comment]
"Backups were configured using backup.velero.io/backup-volumes annotation.
Is there a way to achieve the same(opt-in) behavior with CSI snapshots?"

Opt-in/opt-out switches between fs-backup and snapshots. If you opt out of fs-backup, that means you're using csi snapshots (if you've configured CSI and your volumes are CSI volumes). In other words, if you're using opt-in, then the volumes *not* opted in use CSI snapshots. If your use case is swapped and you want to use fs-backup by default but "opt in" to CSI, then you could just use the opt-out approach. Set `defaultVolumesToFsBackup` to `true` and use `backup.velero.io/backup-volumes-excludes` for the volumes you want to use for CSI.

**Reply by makarov-roman** [reply]
The goal is to be able to tell velero to backup particular PV using annotation. In other words I'm trying to use either CSI or FSB, but never mix them, because It makes no sense in our environment. (e.g. CSI snaphosts disabled when using FSB, node agent disabled when using CSI snapshots)

Opt-in behaviour for our case means, that volumes without annotation are not backed up.

**Reply by sseago** [reply]
"opt-in" and "opt-out" are defined by velero as "opting into fs-backup" and "opting out of fs-backup". These annotations ("backup-volumes" and "backup-volumes-excludes") don't determine whether or not to back up the volume -- they determine whether or not to backup via fs-backup (vs. snapshot). Therefore, "opt-in to fs-backup ('backup-volumes')" is *also* already functioning as "opt-out of snapshot backup", and "opt-out of fs-backup ('backup-volumes-excludes')" is *also* already functioning as "opt-in to snapshot backup".

It sounds like your use case is not swapping between fs-backup and snapshot but rather choosing which volumes to not back up at all. For that use case, the "skip" volume policy would be the way to go, as that is the defined function of this policy. The only problem is that right now you can't define volume policies by label. The proposed label selector volume policy RFE should get you what you want.
